shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,408 @@
+  // Context the proxy function was called from
+  v8::Local<v8::Context> calling_context = args->isolate()->GetCurrentContext();
+  // Context the function was created in
+  v8::Local<v8::Context> func_owning_context =
+      std::get<1>(store->functions()[func_id]).Get(args->isolate());
+
+  v8::Context::Scope func_owning_context_scope(func_owning_context);
+  v8::Local<v8::Function> func =
+      (std::get<0>(store->functions()[func_id])).Get(args->isolate());
+
+  std::vector<v8::Local<v8::Value>> original_args;
+  std::vector<v8::Local<v8::Value>> proxied_args;
+  args->GetRemaining(&original_args);
+  for (auto value : original_args) {
+    proxied_args.push_back(PassValueToOtherContext(
+        calling_context, func_owning_context, value, store));
+  }
+
+  v8::MaybeLocal<v8::Value> maybe_return_value;
+  bool did_error = false;
+  std::string error_message;
+  {
+    v8::TryCatch try_catch(args->isolate());
+    maybe_return_value = func->Call(func_owning_context, func,
+                                    proxied_args.size(), proxied_args.data());
+    if (try_catch.HasCaught()) {
+      did_error = true;
+      auto message = try_catch.Message();
+
+      if (message.IsEmpty() ||
+          !mate::ConvertFromV8(args->isolate(), message->Get(),
+                               &error_message)) {
+        error_message =
+            "An unknown exception occurred in the isolated context, an error "
+            "occurred but a valid exception was not thrown.";
+      }
+    }
+  }
+
+  if (did_error) {
+    v8::Context::Scope calling_context_scope(calling_context);
+    args->ThrowError(error_message);
+    return v8::Local<v8::Object>();
+  }
+
+  if (maybe_return_value.IsEmpty())
+    return v8::Undefined(args->isolate());
+
+  auto return_value = maybe_return_value.ToLocalChecked();
+
+  return PassValueToOtherContext(func_owning_context, calling_context,
+                                 return_value, store);
+}
+
+mate::Dictionary CreateProxyForAPI(mate::Dictionary api,
+                                   v8::Local<v8::Context> source_context,
+                                   v8::Local<v8::Context> target_context,
+                                   RenderFramePersistenceStore* store) {
+  mate::Dictionary proxy =
+      mate::Dictionary::CreateEmpty(target_context->GetIsolate());
+  auto maybe_keys =
+      api.GetHandle()->GetOwnPropertyNames(api.isolate()->GetCurrentContext());
+  if (maybe_keys.IsEmpty())
+    return proxy;
+  auto keys = maybe_keys.ToLocalChecked();
+
+  v8::Context::Scope target_context_scope(target_context);
+  uint32_t length = keys->Length();
+  std::string key_str;
+  int key_int;
+  for (uint32_t i = 0; i < length; i++) {
+    v8::Local<v8::Value> key = keys->Get(target_context, i).ToLocalChecked();
+    // Try get the key as a string
+    if (!mate::ConvertFromV8(api.isolate(), key, &key_str)) {
+      // Try get the key as an int
+      if (!mate::ConvertFromV8(api.isolate(), key, &key_int))
```

Comment:
When are the keys for an object integers?

Reply:
I can't remember the case where I need this, let me test some things and maybe remove it

Reply:
Added a test, double checked we need this.